src/c++/perf_analyzer/load_worker.h

```diff
@@ -182,153 +68,51 @@ class LoadWorker : public IWorker {
-  /// data
-  /// \param valid_inputs The vector of pointers to InferInput objects to be
-  /// used for inference request.
-  /// \param stream_index The data stream to use for next data
-  /// \param step_index The step index to use for next data
-  /// \return cb::Error object indicating success or failure.
-  cb::Error SetInputs(
-      const std::vector<cb::InferInput*>& inputs,
-      std::vector<cb::InferInput*>& valid_inputs, const int stream_index,
-      const int step_index);
+  void SendInferRequest(uint32_t ctx_id, bool delayed = false)
+  {
+    if (on_sequence_model_) {
+      uint32_t seq_stat_index = GetSeqStatIndex(ctx_id);
+      ctxs_[ctx_id]->SendSequenceInferRequest(seq_stat_index, delayed);
+    } else {
+      ctxs_[ctx_id]->SendInferRequest(delayed);
+    }
+  }
 
-  /// Helper function to update the shared memory inputs
-  /// \param inputs The vector of pointers to InferInput objects
-  /// \param stream_index The data stream to use for next data
-  /// \param step_index The step index to use for next data
-  /// \return cb::Error object indicating success or failure.
-  cb::Error SetInputsSharedMemory(
-      const std::vector<cb::InferInput*>& inputs, const int stream_index,
-      const int step_index);
+  virtual std::shared_ptr<InferContext> CreateInferContext()
+  {
+    return std::make_shared<InferContext>(
+        ctxs_.size(), async_, streaming_, on_sequence_model_, using_json_data_,
+        batch_size_, backend_kind_, shared_memory_type_, shared_memory_regions_,
+        start_sequence_id_, sequence_id_range_, sequence_length_, curr_seq_id_,
+        distribution_, thread_stat_, sequence_stat_, data_loader_, parser_,
+        factory_);
+  }
 
- protected:
-  // Create and initialize a new context
+  // Create an inference context and add it to ctxs_
   void CreateContext();
 
-  void PrepAndSendInferRequest(uint32_t ctx_id, bool delayed = false);
-
-  /// A helper function to issue inference request to the server.
-  /// \param context InferContext to use for sending the request.
-  /// \param context_id The ID of the context
-  /// \param request_id The unique id to be associated with the request.
-  /// \param delayed Whether the request fell behind its scheduled time.
-  /// \param callback_func The callback function to use with asynchronous
-  /// request.
-  /// \param async_req_map The map from ongoing request_id to the
-  /// request information needed to correctly interpret the details.
-  /// \param thread_stat The runnning status of the worker thread
-  void SendRequest(
-      std::shared_ptr<InferContext> context, const uint32_t ctx_id,
-      const uint64_t request_id, const bool delayed,
-      cb::OnCompleteFn callback_func,
-      std::map<std::string, AsyncRequestProperties>& async_req_map,
-      std::shared_ptr<ThreadStat> thread_stat);
+  // Any code that needs to execute after the Context has been created
+  virtual void CreateContextFinalize(std::shared_ptr<InferContext> ctx) = 0;
```

Comment:
Why does concurrency worker register async callback but the request rate worker sets active num of threads with this? Why different?

Reply:
Concurrency worker is the only one that has special code that needs to execute at the end of async. RRW's implementation of the function was empty -- there is no need to explicitly register it.

As for the num_threads, this is related to that follow-on story to clean up the tracking of num_threads. RRW knows at time zero how many threads there will be, because it is always max_threads. CW doesn't know until it goes to load-balance the desired concurrency, so that happens later.